[BE] 생성된 유저 데이터 분석 및 조회

b9s2w6 opened the pull request:
생성된 유저들의 데이터를 성별, 연령별, 직업별로 분석하여 무한 스크롤을 활용한 조회 기능 구현

## 변경점
1. **dto**
    - dto 생성 (UserAnalyzeResultResponse, UserSimpleTypeInfo, UserSimpleIdInfo, OccupationNameStats, OccupationCodeStats, NameDto, GenderStats, AgeStats)
    - dto 수정( CreateUserRequestDto Id 값 추가, 로직과 연관된 dto 객체들의 필드값 변경에 맞추어 일부 로직 수정)
2. **DB**
    - User 테이블 컬럼 추가(name, occupation_name, job_number, condition_id)
    - User 테이블 Name 컬럼에  인덱스 추가 (1만건 데이터 조회 기준 681ms -> 465ms 조회속도 향상)
3. **랜덤이름 생성 기능 구현**
    - 나이와 성별을 파라미터로 넘긴다면 해당 파라미터에 적합한 이름을 랜덤으로 반환
4. **생성된 랜덤 데이터 분석 기능 구현**
    - 성별, 연령별, 직업별 분포 인원 분석
5. **NameUtil 기능 세분화**
    - Name cache init 로직과 이름 생성 로직 분할
6. **Simulation Controller 및 서비스 삭제**
7. **복수형 클래스 이름 단수형 이름으로 수정**
 
## TODO
- 나이와 성별에 해당하는 이름 풀이 적어서 중복된 이름 다수 발생함으로 개선 필요.

## 체크리스트
- [x] 통합테스트 완료
- [x] 테스트 코드 작성

## 관련 이슈
- Closes #26

b9s2w6 committed:
#26 - feat: 사용자 생성 시 랜덤이름 발생 기능 추가

서비스가 시작되면 json에 들어있는 이름들을 캐시에 올리고 클라이언트를 생성할 때, nameUtil을 통해 이름을 랜덤으로 제공한다.

b9s2w6 committed:
#26 - feat : 사용자 랜덤 생성 API 구현 및 랜덤 이름 생성 테스트 코드 작성

사용자를 랜덤 생성하고 이를 DB에 저장함.

b9s2w6 committed:
#26 - feat: 사용자 생성, 조회, 삭제, 분석 기능 추가

랜덤으로 유저를 생성하고 이를 성별, 연령별, 직업별로 분석한 정보를 사용자에게 제공한다.

b9s2w6 committed:
Merge branch 'develop' of https://github.com/SIMPAYLOG/finance-data-generator into feat/26-user-analyze

KHJHK commented on generator-core/src/main/java/com/simpaylog/generatorcore/dto/NameDto.java:
```diff
@@ -0,0 +1,22 @@
+    @JsonProperty("Age Group")
+    private int ageGroup;
+
+    public String getFullName() { return fullName; }
```
@Getter쓰면 더 깔끔해 보일 것 같습니다 ;)

KHJHK commented on generator-core/src/main/java/com/simpaylog/generatorcore/utils/NameUtil.java:
```diff
@@ -0,0 +1,58 @@
+                        .add(user.getFullName());
+            }
+
+            System.out.println("JSON 데이터 로딩 완료.");
```
민우 cache 코드 보면, 캐시 작업 완료/실패 후 로그 양식을 통일해놨어요. 이부분도 캐시 로딩 부분이니까 로그 양식 통일하면 더 가독성이 좋을 것 같아요.

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/dto/request/UserGenerationConditionRequestDto.java:
```diff
@@ -19,8 +19,9 @@ public record UserGenerationConditionRequestDto(
         String occupationCode
 ) {
-    public static UserGenerationCondition toCore(UserGenerationConditionRequestDto dto) {
+    public static UserGenerationCondition toCore(UserGenerationConditionRequestDto dto, int id) {
```
toCore 보다 toUserGenerationCondition 으로 메서드 명을 바꾸는 게 좋겠다!

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/controller/UserController.java:
```diff
@@ -0,0 +1,65 @@
+
+    @GetMapping("/analyze")
+    public Response<UserAnalyzeResultResponse> analyzeUsers() {
+        UserAnalyzeResultResponse userAnalyzeResultResponse
+                = userService.analyzeUsers();
```
변수로 안빼고 리턴해줘도 좋을 것 같아요.
잘못됐다는 건 아닙니다!!ㅎㅎ

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/dto/AgeStats.java:
```diff
@@ -0,0 +1,4 @@
+package com.simpaylog.generatorcore.dto;
+
+public record AgeStats(Integer ageGroup, Long count) {
```
클래스 이름 단수명으로 가도 좋을 듯

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/dto/GenderStats.java:
```diff
@@ -0,0 +1,4 @@
+package com.simpaylog.generatorcore.dto;
+
+public record GenderStats(Integer male, Integer female) {
```
클래스 이름 단수명으로 가도 좋을 듯

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/dto/OccupationCodeStats.java:
```diff
@@ -0,0 +1,7 @@
+package com.simpaylog.generatorcore.dto;
+
+public record OccupationCodeStats(
```
단수명!

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/service/UserService.java:
```diff
@@ -28,7 +36,58 @@ public void deleteUserAll() {
+    public Page<UserSimpleTypeInfo> findUsersByPage(Pageable pageable) {
+        Page<User> userPage = userRepository.findAllByOrderByNameAsc(pageable);
+
+        return userPage.map(user -> new UserSimpleTypeInfo(
+                user.getName(),
+                user.getGender(),
+                user.getAge(),
+                preferenceLocalCache.get(user.getUserBehaviorProfile().getPreferenceId()).name(),
+                user.getOccupationName()
```
UserSimpleTypeInfo에 정적 팩토리 메서드 추가하고, 활용한다면 코드가 더 짧아질 수 있어요.

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/utils/NameUtil.java:
유저 이름 중복되는 경우가 있으므로 주석으로 // TODO 걸어주세요.

lmw7414 reviewed (changes requested):
수정하고 다시 전체 테스트 확인해주세요

b9s2w6 committed:
#26 - fix : Simulation Conroller 및 Service 제거, 다수 엔티티 이름 변경 및 필드 수정

1. 복수형 이름 다수형으로 변경
- ageStats -> ageStat
- genderStats -> genderStrat
- occupaionCodeStats -> occupaionCodeStat
- occupaionNameStats -> occupaionNameStat

2. NameUtil 캐시 생성 로직과 이름 생성 로직 분리

3. Simulation Conroller 및 Service 제거

4. UserContoller 및 서비스 불필요 메서드 제거 및 수정

b9s2w6 committed:
#26 - fix : NameInfo 자료형 변경 및 유저 생성 수정

1. NameInfo 자료형 레코드로 수정
2. 생성된 User DB 저장 메소드 누락으로 인한 추가

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/cache/UserNameLocalCache.java:
log.info에 내용 구체적으로 변경해주세요. 다른 캐시 클래스 참조!

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/service/UserService.java:
```diff
@@ -28,7 +44,48 @@ public void deleteUserAll() {
+        return result;
+    }
+
+    List<UserInfoResponse> perferenceIdtoType(List<UserInfoDto> users) {
+        return users.stream()
+                .map(user -> new UserInfoResponse(
+                        user.name(),
+                        user.gender(),
+                        user.age(),
+                        preferenceLocalCache.get(user.preferenceId()).name(),
+                        user.occupationName()
+                ))
+                .collect(Collectors.toList());
+    }
+
+    List<OccupationNameStat> occupationCodeToName(List<OccupationCodeStat> occupationCodeStats) {
+        return occupationCodeStats.stream()
+                .map(stat -> new OccupationNameStat(
+                        occupationLocalCache.get(stat.occupationCategory()).occupationCategory(),
+                        stat.count()
+                ))
+                .collect(Collectors.toList());
+    }
```
접근 제어자 명시해주시고, 메서드명이 이해하기 힘듭니다. 주석이든 메서드를 구체적으로든 변경해주시면 좋을 것 같습니다.

lmw7414 commented on generator-core/src/test/java/com/simpaylog/generatorcore/utils/NameGeneratorTest.java:
NameUtil 클래스와 이름 통일 시켜주세요.

lmw7414 commented:
수정 사항 변경 후 전체 테스트 이상 없으며 merge해도 좋습니다.

b9s2w6 committed:
#26 - fix : 코드 리팩토링

1. userService 메소드 주석 추가
2. 로그 메시지 구체화
3. nameInfo 불필요 어노테이션 제거
4. names.json 키 값 파스칼 형태로 수정
